mysql/mysqldump_test.go

```diff
@@ -0,0 +1,561 @@
+				"INSERT INTO test (a, b, c, d, e, f) VALUES (NULL,NULL,NULL,NULL,NULL,42,6);\n", // Error
+			expectedData: []spannerData{
+				spannerData{
+					table: "test", cols: []string{"a", "b", "c", "d", "e", "f", "synth_id"},
+					vals: []interface{}{int64(7), float64(42.1), true,
+						getDate("2019-10-29"), []byte{0x89, 0x50},
+						[]spanner.NullString{{StringVal: "42", Valid: true}, {StringVal: "6", Valid: true}},
+						bitReverse(0)}},
+				spannerData{table: "test", cols: []string{"a", "synth_id"}, vals: []interface{}{int64(7), bitReverse(1)}},
+				spannerData{table: "test", cols: []string{"b", "synth_id"}, vals: []interface{}{float64(42.1), bitReverse(2)}},
+				spannerData{table: "test", cols: []string{"c", "synth_id"}, vals: []interface{}{true, bitReverse(3)}},
+				spannerData{table: "test", cols: []string{"d", "synth_id"}, vals: []interface{}{getDate("2019-10-29"), bitReverse(4)}},
+				spannerData{table: "test", cols: []string{"e", "synth_id"}, vals: []interface{}{[]byte{0x89, 0x50}, bitReverse(5)}},
+				spannerData{table: "test", cols: []string{"f", "synth_id"},
+					vals: []interface{}{[]spanner.NullString{{StringVal: "42", Valid: true}, {StringVal: "6", Valid: true}}, bitReverse(6)}},
+			},
+		},
+	}
+	for _, tc := range dataErrorTests {
+		conv, rows := runProcessMySQLDump(tc.input)
+		assert.Equal(t, tc.expectedData, rows, tc.name+": Data rows did not match")
+		assert.Equal(t, conv.BadRows(), int64(5), tc.name+": Error count did not match")
+	}
+}
+
+// The following test Conv API calls based on data generated by ProcessMySQLDump.
+func TestProcessMySQLDump_GetDDL(t *testing.T) {
+	conv, _ := runProcessMySQLDump("CREATE TABLE cart (productid text, userid text, quantity bigint);\n" +
+		"ALTER TABLE cart ADD CONSTRAINT cart_pkey PRIMARY KEY (productid, userid);")
+	expected := "CREATE TABLE cart (\n" +
+		"productid STRING(MAX) NOT NULL,\n" +
+		"userid STRING(MAX) NOT NULL,\n" +
+		"quantity INT64\n" +
+		") PRIMARY KEY (productid, userid)"
+	// normalizeSpace isn't perfect, but it handles most of the
+	// usual discretionary space issues.
+	c := ddl.Config{}
+	assert.Equal(t, normalizeSpace(expected), normalizeSpace(strings.Join(conv.GetDDL(c), " ")))
+}
+
+func TestProcessMySQLDump_Rows(t *testing.T) {
+	conv, _ := runProcessMySQLDump("CREATE TABLE cart (a text, n bigint);\n" +
+		"INSERT INTO cart (a, n) VALUES ('a42', 2);")
+	assert.Equal(t, int64(1), conv.Rows())
+}
+
+func TestProcessMySQLDump_BadRows(t *testing.T) {
+	conv, _ := runProcessMySQLDump("CREATE TABLE cart (a text, n bigint);\n" +
+		"INSERT INTO cart (a, n) VALUES ('a42', 'not_a_number');")
+	assert.Equal(t, int64(1), conv.BadRows())
+}
+
+func TestProcessMySQLDump_GetBadRows(t *testing.T) {
+	conv, _ := runProcessMySQLDump("CREATE TABLE cart (a text, n bigint);\n" +
+		"INSERT INTO cart (a, n) VALUES ('a42', 'not_a_number');")
+	assert.Equal(t, 1, len(conv.SampleBadRows(100)))
+}
+
+func TestProcessMySQLDump_AddPrimaryKeys(t *testing.T) {
```

Comment:
consider to use `t.Run()` (https://github.com/golang/go/wiki/TableDrivenTests#example-of-a-table-driven-test).

Reply:
Done. Added `t.Run()` in `mysqldump_test.go` and `data_test.go`